app/src/main/java/de/westnordost/streetcomplete/quests/LastPickedValuesStore.kt

```diff
@@ -30,29 +30,37 @@ class LastPickedValuesStore<T : Any>(
 
 private const val MAX_ENTRIES = 100
 
-/* In the first `historyCount` items, return the `count` most-common non-null items, in order.
- * If the first item is not included (and is not null), it replaces the last of the common items.
- * If fewer than `count` unique items are found, continue counting items until that many are found,
- * or the end of the sequence is reached.
+/* Returns the `count` most-common non-null items in the first `historyCount`
+ *  items of the sequence, in their original order.
+ * If there are fewer than `count` unique items, continues counting items
+ *  until that many are found, or the end of the sequence is reached.
+ * If the first item is not null, it is always included, displacing the
+ *  least-common of the other items if necessary.
  */
 fun <T : Any> Sequence<T?>.mostCommonWithin(count: Int, historyCount: Int): Sequence<T> {
     val counts = this.countUniqueNonNull(historyCount, count)
-    val top = counts.keys.sortedByDescending { counts.get(it) }
+    val top = counts.keys.sortedByDescending { counts[it]!!.count }.take(count)
     val latest = this.take(1).filterNotNull()
     val items = (latest + top).distinct().take(count)
-    return items.sortedByDescending { counts.get(it) }
+    return items.sortedBy { counts[it]!!.indexOfFirst }
 }
 
+private data class ItemStats(val indexOfFirst: Int, var count: Int = 0)
+
 // Counts at least the first `minItems`, keeps going until it finds at least `target` unique values
-private fun <T : Any> Sequence<T?>.countUniqueNonNull(minItems: Int, target: Int): Map<T, Int> {
-    val counts = mutableMapOf<T, Int>()
-    val items = takeAtLeastWhile(minItems) { counts.size < target }.filterNotNull()
-    return items.groupingBy { it }.eachCountTo(counts)
-}
+private fun <T : Any> Sequence<T?>.countUniqueNonNull(minItems: Int, target: Int): Map<T, ItemStats> {
+    val counts = mutableMapOf<T, ItemStats>()
 
-// Take at least `count` elements, then continue until `predicate` returns false
-private fun <T> Sequence<T>.takeAtLeastWhile(count: Int, predicate: (T) -> Boolean): Sequence<T> =
-    withIndex().takeWhile{ (i, t) -> i < count || predicate(t) }.map { it.value }
+    // Sequences are evaluated lazily, so the `forEach` affects each `counts.size` check
```

Comment:
I don't get this comment. Does `forEach` cause all items to be evaluated? Then I'd suggest:

```suggestion
    // Sequences are evaluated lazily, but `forEach` causes all items to be evaluated
```

Reply:
Does the `filter` (which has no special interaction with sequences here) get in the way of that point being obvious? I could pull it into the `forEach`
```kt
    this.withIndex()
        .takeWhile { it.index < minToCount || counts.size < target }
        .forEach {
            if (it.value != null) {
                counts.getOrPut(it.value!!) { ItemStats(it.index) }.count++
            }
        }
```
Or up onto the previous line
```kt
    this.withIndex().filter { it.value != null }
        .takeWhile { it.index < minItems || counts.size < target }
        .forEach {
            counts.getOrPut(it.value!!) { ItemStats(it.index) }.count++
        }
```

Reply:
I updated the comment. If the new one does not clarify well enough or you like either of those formattings, you can un-resolve this or add a new comment :)